### DISCUSSION TITLE
How can Shadow task scheduling support blocking operation?

### DISCUSSION THREAD

**Discussion by xiex386** [shadow/shadow#3416] [Q&A]
I'm working on integrate Linux networking stack (a userspace Linux kernel) into Shadow. But when shadow delegates some blocking syscall (e.g connect() ) to the networking stack, the latter will schedule out its worker thread and prevent Shadow from dealing with the following events that might resume the execution, thus dead lock.  I understand Shadow handles blocking syscall by returning special value immediately, adding it to event queue, and restarting it when possible. So it might be strange to do this work. But I can't come up with some ideas. Could anyone help me?

**Comment by sporksmith** [comment]
I'm not following exactly what you're trying to do. I think it'd help if you gave a little more context about what you're trying to accomplish. Are you trying to allow shadow processes to talk to the real network via this stack? Replace shadow's model of the simulated network with a different model? Something else?

**Reply by xiex386** [reply]
Thanks for your reply.
Shadow's simulated network stack takes great effort and I appreciate the time and dedication you’ve invested. But it still lack some real behavior, like multi-hop forwarding or TCP congestion control, and it avoid these by simplifying the network topology, right? For application example,  some routing protocol apps like `bird` failed to run because it can't configure the routing.
So I'm trying to replace shadow's simulated network with a more realistic one. Since shadow is designed to support native Linux app, why not use Linux networking stack directly? With this idea, I found some userspace linux networking stack ([NUSE](https://github.com/libos-nuse/net-next-nuse)) or complete kernel ([lkl](https://github.com/lkl/linux)) and I try to replace shadow's networking syscall handler with them. But some blocking syscall like connect() in these handler will schedule out the current thread (shadow worker thread), so shadow can't execute next events on the same host, which might wake up the worker thread. When this happens, shadow will wait indefinitely.
The above was my initial problem. Now I come up with a possible workaround. Maybe I could add `NONBLOCK` flags to these hanlder calls whatever the original user syscall flags are, and decide whether return to user or add to shadow's event queue according to user's `NONBLOCK` flag.

**Reply by stevenengler** [reply]
> Since shadow is designed to support native Linux app, why not use Linux networking stack directly? With this idea, I found some userspace linux networking stack (NUSE) or complete kernel (lkl) and I try to replace shadow's networking syscall handler with them.

This sounds really cool and I hope it works out!

> Shadow's simulated network stack takes great effort and I appreciate the time and dedication you’ve invested. But it still lack some real behavior, like multi-hop forwarding or TCP congestion control, and it avoid these by simplifying the network topology, right?

Yeah, Shadow's network stack is missing things that we'd really like to have, but it's complicated and a lot of work. We have started rewriting our network stack (starting with the tcp code), but none of us have a lot of time to work on it at the moment. Another thing to keep in mind is that we want it to be fast and not have a lot of overhead (CPU, memory, etc). Some network stacks are designed assuming that they'll be the only network stack running on the machine, but in Shadow we need to support scaling to run tens of thousands of network stacks (one for each host).

> But some blocking syscall like connect() in these handler will schedule out the current thread (shadow worker thread), so shadow can't execute next events on the same host, which might wake up the worker thread. When this happens, shadow will wait indefinitely.

This part isn't 100% clear to me. It sounds like you're calling the new network stack's `connect` function from Shadow's `connect` system call handler? And the network stack's function is blocking? In this case, yeah, you cannot call blocking functions from within Shadow's system call handlers. In Shadow, the system call handler like `connect` needs to return something immediately and would return a `SyscallError::Blocked(condition)` in this case. Then Shadow can watch for the condition to become unblocked so it can restart the `connect` system call handler and unblock the managed application's thread.

I think your two options would be to:

1. Run the blocking methods in another Shadow thread, and then extend Shadow's `SyscallCondition` to be able to wake the simulated thread when the blocking network stack's function returns in that other thread. This sounds like it could be difficult, and would create a lot of extra threads in your simulation.
2. What you mentioned above where you call into the network stack with a NONBLOCK flag so that it never blocks, and simulate the blocking behaviour from within Shadow. This is pretty much [what Shadow does now](https://github.com/shadow/shadow/blob/86c19651c9471dff52a944787f74bc0aa754bd79/src/main/host/descriptor/socket/inet/legacy_tcp.rs#L821) with its own network stack. The networking implementation itself never blocks and will always return `EWOULDBLOCK`. Then Shadow checks the socket flag to decide whether to return either `EWOULDBLOCK` or `SyscallError::Blocked(condition)` from the system call handler.
